src/main/cpp/src/shuffle_split_detail.hpp

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/utilities/traits.hpp>
+
+namespace spark_rapids_jni {
+
+namespace detail {
+
+/**
+ * @brief Functor for computing size of data elements for a given cudf type.
+ *
+ * Note: columns types which themselves inherently have no data (strings, lists,
+ * structs) return 0.
+ */
+struct size_of_helper {
+  template <typename T>
+  constexpr std::enable_if_t<!cudf::is_fixed_width<T>() && !std::is_same_v<T, cudf::string_view>,
+                             size_t>
+  operator()() const
+  {
+    return 0;
+  }
+
+  template <typename T>
+  constexpr std::enable_if_t<!cudf::is_fixed_width<T>() && std::is_same_v<T, cudf::string_view>,
+                             size_t>
+  operator()() const
+  {
+    return sizeof(cudf::device_storage_type_t<int8_t>);
+  }
+
+  template <typename T>
+  constexpr std::enable_if_t<cudf::is_fixed_width<T>(), size_t> __device__
+  operator()() const noexcept
+  {
+    return sizeof(cudf::device_storage_type_t<T>);
+  }
+};
+
+/**
+ * @brief Header for each partition.
+ */
+struct partition_header {
+  uint32_t magic_number;
+  uint32_t version;
+  uint64_t offset;
+  uint64_t num_rows;
+  uint64_t validity_size;
+  uint64_t offset_size;
+  uint64_t data_size;
+};
```

Comment:
We appear to be missing a `hasValidityBuffer` entry at the end here. I know it is variable length (based on the number of columns), so is it just not returned to the user, or is it not a part of this code base yet?

https://github.com/NVIDIA/spark-rapids-jni/blob/74e33d948442d72cd45dee2f3e062603d4953284/src/main/java/com/nvidia/spark/rapids/jni/kudo/KudoSerializer.java#L112-L124

Reply:
It's in there, it's just not part of the fixed-sized part of the header (the `partition_header` struct).

https://github.com/NVIDIA/spark-rapids-jni/blob/8a8e920f771529f9d353f03fdcd3c307adc4aac9/src/main/cpp/src/shuffle_split.cu#L642